scripts/cloudbuild.py

```diff
@@ -0,0 +1,179 @@
+flags.DEFINE_string('config', None, 'input configuration file')
+flags.DEFINE_string('base_url', _BASE_URL, 'base url for images')
+flags.DEFINE_integer('timeout_sec', 24 * 60 * 60,
+                     'timeout for all builds to complete')
+
+# ----------------------------------------------------------------------------
+logging.get_absl_handler().setFormatter(None)
+
+Config = Dict[str, Any]
+
+
+# ----------------------------------------------------------------------------
+class BuildStep(NamedTuple):
+  '''a google cloud-build step'''
+  sid: str
+  name: str = _CLOUD_BUILDER
+  args: List[str] = []
+  dependencies: List[str] = ['-']
+
+  def to_dict(self) -> Dict[str, Any]:
+    '''converts step to dict'''
+    return {
+        'id': self.sid,
+        'name': self.name,
+        'args': self.args,
+        'waitFor': self.dependencies,
+    }
+
+
+# ----------------------------------------------------------------------------
+class ImageSpec(NamedTuple):
+  '''a caliban base image spec'''
+  dockerfile: str = _CPU_DOCKERFILE
+  base_image: Optional[str] = _CPU_BASE_IMAGE
+  cpu_version: Optional[str] = None
+  gpu_version: Optional[str] = None
+  build_args: List[str] = []
+
+  def tag(self) -> str:
+    '''returns tag for this spec'''
+    parts = []
+    if self.gpu_version is None:  # cpu image
+      parts = [_CPU_TAG, self.cpu_version]
+    else:  # gpu image
+      parts.append(_GPU_TAG)
+      if self.cpu_version is None:  # gpu base image
+        parts += ['base', self.gpu_version]
+      else:  # full gpu image
+        parts += [self.gpu_version, self.cpu_version]
+
+    return f'{FLAGS.base_url}:' + '-'.join(parts)
+
+
+# ----------------------------------------------------------------------------
+def _create_push_step(tag: str) -> BuildStep:
+  '''creates a cloudbuild push step'''
+  return BuildStep(sid=f'{tag}-push', args=['push', tag], dependencies=[tag])
+
+
+# ----------------------------------------------------------------------------
+def _create_gpu_base_image_specs(cfg: Config) -> Dict[str, ImageSpec]:
+  '''creates gpu base image specs'''
+  return {
+      k: ImageSpec(base_image=None,
+                   gpu_version=k,
+                   build_args=v,
+                   dockerfile=_GPU_DOCKERFILE) for k, v in cfg.items()
+  }
+
+
+# ----------------------------------------------------------------------------
+def _create_cpu_image_specs(cfg: Config) -> Dict[str, ImageSpec]:
+  '''creates cpu image specs'''
+  return {k: ImageSpec(cpu_version=k, build_args=v) for k, v in cfg.items()}
+
+
+# ----------------------------------------------------------------------------
+def _create_image_spec(
+    cfg: Config,
+    cpu_specs: Dict[str, ImageSpec],
+    gpu_specs: Dict[str, ImageSpec],
+) -> ImageSpec:
+  '''creates an image spec from a config dictionary and gpu/gpu specs'''
+  gpu_version = cfg.get('gpu')
+  cpu_spec = cpu_specs[cfg.get('cpu')]
+
+  if gpu_version is None:
```

Comment:
the mode switcher would help clean this up too.

Reply:
Can't use that here, but modified other calls to use this mechanism.